### PR TITLE
RouteOptions applier: avoid overwritten values

### LINKED ISSUE #4174: MapboxOffboardRouter and MapboxOnboardRouter replace pre-set route options
follow up https://github.com/mapbox/mapbox-navigation-android/pull/4171#discussion_r600723964

Investigate how properties substitute each other (`fun RouteBuilderProvider.getBuilder()` by `fun MapboxDirections.Builder.routeOptions()`)

Applicable for `MapboxOffboardRouter` and `MapboxOnboardRouter`

https://github.com/mapbox/mapbox-navigation-android/blob/07fe4ba304cad3438797ea968932a2f9ec1fff21/libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt#L48-L55

### PR DESCRIPTION
**Pull request by RingerJK**
### Description

RouteOptions applier: avoid overwritten values
Closes #4174 

### Changelog
```
<changelog>
:warning: Requesting a route via `MapboxNavigation#requestRoutes` does not automatically append any defaults. Defaults are now available under these extensions:
- `RouteOptions.Builder.applyDefaultNavigationOptions()` that applies the options that are required for the route request to execute or otherwise recommended for the Navigation SDK and all of its features to provide the best car navigation experience.
- `RouteOptions.Builder.applyLanguageAndVoiceUnitOptions(context: Context)` that applies the options that adapt the returned instructions' language and voice unit based on the device's `Locale`.

The extensions are very much recommended to be called when building the `RouteOptions` object and then tweaked for a specific use case.
</changelog>
```

### CODE REVIEW

**Comment by LukasPaczos** [comment]
Why not consolidate `RouteBuilderProvider#getBuilder` and `MapboxDirections.Builder.routeOptions` into a single method? Both apply some changes to the builder, often modifying the same fields.

**Comment by RingerJK** [comment]
> Why not consolidate `RouteBuilderProvider#getBuilder` and `MapboxDirections.Builder.routeOptions` into a single method? Both apply some changes to the builder, often modifying the same fields.

I thought about it and rejected to keep `getBuilder` to have a single source of default values were easy to check which are default(in case if we move it to single method default values will be mixed with set values). 
The current design separates providing default values and applying RouteOptions which I like actually

**Comment by LukasPaczos** [comment]
Gotcha, I think we could work on naming the methods a little bit, in this case, to make this distinction clear.

Another issue I'd love to see resolved as part of this contribution is the ambiguity of what Nav SDK actually requests. Now, for example, we apply the
```
.annotations(
    listOf(
        DirectionsCriteria.ANNOTATION_CONGESTION,
        DirectionsCriteria.ANNOTATION_DISTANCE
    )
)
```
by default, and there is no way for the developer to clear it. They create the `RouteOptions`, but if those are do not include `annotations`, we'll always add them anyway. What if developers actually didn't want to receive `annotations` to save on bandwidth?

I'd rather have an entity that could help developers provide a set of default options than us filling them up under the hood. I think this has been a major source of confusion in the past.

What if `MapboxNavigation` would expose different overloads, like:
```
// what you pass is exactly what we request
fun requestRoutes(
    routeOptions: RouteOptions,
    routesRequestCallback: RoutesRequestCallback
): Long

// you can modify the default set of options (the ones that we produce under the hood currently)
fun requestRoutes(
    routesRequestCallback: RoutesRequestCallback,
    block: (RouteOptions.Builder) -> Unit
): Long

// get default options ahead of time and pass them to `MapboxNavigation#requestRoutes`
RouteOptionsProvider.getDefault()
```

What do you think @RingerJK?

**Comment by RingerJK** [comment]
> by default, and there is no way for the developer to clear it. 
that's a great point, I feel we should be good to investigate it in the current PR 👍 

I'd expose only one way at least in the first iteration. Let me try

File: examples/src/main/java/com/mapbox/navigation/examples/util/RouteOptionsEx.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+@file:JvmName("RouteOptionsEx")
+package com.mapbox.navigation.examples.util
+
+import android.content.Context
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.extensions.applyDefaultOptions
+import com.mapbox.navigation.base.extensions.applyLocationAndVoiceUnit
+import com.mapbox.navigation.base.extensions.applyRecommendedOptions
+
+fun RouteOptions.Builder.applyAllOptions(context: Context) = apply {
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/util/RouteOptionsEx.kt:10]: Should we also add this extension function into `RouteOptionsEx`?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt [hunk 2]
```diff
@@ -36,3 +40,44 @@ fun RouteOptions?.supportsRouteRefresh(): Boolean {
...
     return isTrafficProfile && isOverviewFull && hasCongestionOrMaxSpeed
 }
+
+/**
+ * Applies default [RouteOptions] parameters to the RouteOptions builder
+ *
+ * @receiver RouteOptions.Builder
+ * @return RouteOptions.Builder
+ */
+fun RouteOptions.Builder.applyDefaultOptions(): RouteOptions.Builder = also {
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt:50]: Maybe? 🤔 

```suggestion
fun RouteOptions.Builder.applyDefaultRequiredOptions(): RouteOptions.Builder = also {
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt [hunk 3]
```diff
@@ -36,3 +40,44 @@ fun RouteOptions?.supportsRouteRefresh(): Boolean {
...
+    user(Constants.MAPBOX_USER)
+    profile(DirectionsCriteria.PROFILE_DRIVING)
+    geometries(DirectionsCriteria.GEOMETRY_POLYLINE6)
+    requestUuid("")
+}
+
+/**
+ * Apply [RouteOptions] location and voice unit based on [Context]
+ */
+fun RouteOptions.Builder.applyLocationAndVoiceUnit(context: Context): RouteOptions.Builder = also {
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt:61]:
```suggestion
fun RouteOptions.Builder.applyLanguageAndVoiceUnitFrom(context: Context): RouteOptions.Builder = also {
```

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/MapboxDirectionsUtils.kt [hunk 4]
```diff
@@ -31,7 +34,9 @@ internal fun MapboxDirections.Builder.routeOptions(
         }
     }
 
-    alternatives(options.alternatives() ?: false)
```
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/MapboxDirectionsUtils.kt:34]: Removing these we're not opinionated anymore (unless user explicitly calls `applyRecommendedOptions`). These options are `null` by default and we're not falling back (`?:`) i.e. these options only change when they're provided beforehand. Is that what we want? 🤔 

The same thing applies to the other removals 👇

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt (hunk 2 above)
Comment (Guardiola31337) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt:50]: These are required options, should we add them implicitly downstream? Clients would need to explicitly call this to avoid `RouteOptions` `IllegalArgumentException`s and we should probably default to these internally to prevent these crashes and not force developers to remember calling it.

**Review by Guardiola31337 — COMMENTED** [review]
Left some comments to discuss before merging here.

File: examples/src/main/java/com/mapbox/navigation/examples/util/RouteOptionsEx.kt (hunk 1 above)
Comment (LukasPaczos) [reply examples/src/main/java/com/mapbox/navigation/examples/util/RouteOptionsEx.kt]: Opted to remove it instead.

**Comment by LukasPaczos** [comment]
This PR now introduces 2 new extensions:
- `RouteOptions.Builder.applyDefaultNavigationOptions()` that applies the options that are required for the route request to execute or otherwise recommended for the Navigation SDK and all of its features to provide the best car navigation experience.
- `RouteOptions.Builder.applyLanguageAndVoiceUnitOptions(context: Context)` that applies the options that adapt the returned instructions' language and voice unit based on the device's `Locale`.

I tried to be as clear as possible in the documentation that these options are required and very much recommended to set, while also avoiding undefined and surprising behavior that was previously caused by us setting these values under the hood.

Ideally, those should be the defaults that are coming from `mapbox-java` itself but on the other hand, `mapbox-java` is generic, it can be used standalone so applying these default options might've been surprising as well, that's why I opted to expose extensions here instead.

Out of curiosity, how is iOS Nav SDK or `mapbox-directions-swift` handling any defaults for route requests? Are they set by default by `mapbox-directions-swift`, set under the hood by the Nav SDK, or just documented to apply by developers?

**Comment by LukasPaczos** [comment]
cc @mapbox/navigation-ios ^^^

**Comment by Guardiola31337** [comment]
Noting that `unit-tests` are ❌ 

```
java.lang.ClassNotFoundException: com.mapbox.navigation.base.extensions.RouteOptionsEx
```

File: scripts/validate-changelog.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
         skip_changelog = True
         break
 
-  matches = re.search(r'<changelog>(.+)</changelog>', pr_description) is not None
+  matches = re.search(r'<changelog>(.+)</changelog>', pr_description, flags=re.S) is not None
```
Comment (Guardiola31337) [inline scripts/validate-changelog.py:25]: Is this related to OP or some cleanup? 🤔

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/RouteOptionsEx.kt [hunk 6]
```diff
@@ -9,7 +9,7 @@ import com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl
  * - uuid is not empty;
  * - uuid is not equal to [MapboxNativeNavigatorImpl.OFFLINE_UUID].
  */
-fun RouteOptions?.isUuidValidForRefresh(): Boolean =
+internal fun RouteOptions?.isUuidValidForRefresh(): Boolean =
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/RouteOptionsEx.kt:12]: If this is `internal` now, it can be moved out of the `internal` package.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/utils/RouteOptionsInternalExKtTest.kt [hunk 7]
```diff
@@ -2,11 +2,11 @@ package com.mapbox.navigation.core.internal.utils
 
 import com.mapbox.api.directions.v5.models.RouteOptions
 import com.mapbox.geojson.Point
-import com.mapbox.navigation.base.internal.extensions.applyDefaultParams
+import com.mapbox.navigation.base.extensions.applyDefaultNavigationOptions
 import org.junit.Assert
 import org.junit.Test
 
-class RouteOptionsExTest {
+class RouteOptionsInternalExKtTest {
```
Comment (Guardiola31337) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/internal/utils/RouteOptionsInternalExKtTest.kt:9]:
```suggestion
class RouteOptionsInternalExTest {
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt [hunk 8]
```diff
@@ -32,7 +38,45 @@ fun RouteOptions?.supportsRouteRefresh(): Boolean {
...
+            it == DirectionsCriteria.ANNOTATION_CLOSURE
     } ?: false
     return isTrafficProfile && isOverviewFull && hasCongestionOrMaxSpeed
 }
+
+/**
+ * Applies the [RouteOptions] that are required for the route request to execute
+ * or otherwise recommended for the Navigation SDK and all of its features to provide the best car navigation experience.
+ */
+fun RouteOptions.Builder.applyDefaultNavigationOptions(): RouteOptions.Builder = apply {
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt:51]: This https://github.com/mapbox/mapbox-navigation-android/pull/4320#discussion_r626616826 still applies. What do you think?

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/MapboxDirectionsUtils.kt (hunk 4 above)
Comment (Guardiola31337) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/MapboxDirectionsUtils.kt:34]: This https://github.com/mapbox/mapbox-navigation-android/pull/4320#discussion_r626612065 still applies. What do you think?

**Review by Guardiola31337 — COMMENTED** [review]
Left some comments @LukasPaczos

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt (hunk 8 above)
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt:51]: My thoughts are in https://github.com/mapbox/mapbox-navigation-android/pull/4320#issuecomment-839710568. I hope that we can be transparent about it and document it properly instead of changing values under the hood and introducing surprising behavior.

File: scripts/validate-changelog.py (hunk 5 above)
Comment (LukasPaczos) [reply scripts/validate-changelog.py:25]: Unrelated, allows for changelog message tag with new lines.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt (hunk 8 above)
Comment (Guardiola31337) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/RouteOptionsEx.kt:51]: Ohh I meant only for the required options.

**Review by Guardiola31337 — APPROVED** [review]
